### ISSUE TITLE
Performance Concern about CrossEntropyLoss

### ISSUE THREAD

**Issue by JubilantJerry** [pytorch/pytorch#119261]
### 🐛 Describe the bug

I've seem a [claim](https://unsloth.ai/blog/mistral-benchmark) floating around that PyTorch's CrossEntropyLoss operation is poorly optimized during the backward pass, and that it can be rewritten to reduce VRAM usage by a massive amount (seemingly by 3x or more!), with no impact on computation speed.

This sounds like a tall claim given CrossEntropyLoss is very simple operation, but the authors of the claim has created a [repo](https://github.com/unslothai/unsloth/tree/main) which contains, among other optimizations, an alternate CUDA kernel for [CrossEntropyLoss](https://github.com/unslothai/unsloth/blob/main/unsloth/kernels/cross_entropy_loss.py) using Triton, and some people have indeed observed significant VRAM improvement using the author's repo. My understanding is that PyTorch compiles its own CUDA kernel for CrossEntropyLoss, so there doesn't seem to be any reason why PyTorch could not implement it in the most optimal way.

I'm not so sure if I should've marked the issue as a bug since I'm not certain of the veracity of this performance optimization claim, but if it's true this seems like a very big performance bug in PyTorch's CrossEntropyLoss implementation. Perhaps someone could take a look to see if it's plausible to optimize PyTorch's CrossEntropyLoss in a similar way? Given that LLM's have a very large logits tensor due to their vocabulary size, an optimization of this operation can potentially speed up the training of a huge number of workloads.

### Versions

PyTorch 2.1.1

**Comment by wanchaol** [comment]
cc @tianyu-l I think this is probably related to the nll_loss_backward + softmax_backward discussion we had yesterday

**Comment by wanchaol** [comment]
@albanD any core tags we can add to this issue and triage? I feel this is something we should investigate into so don't want this issue to get lost

**Comment by albanD** [comment]
high priority to see how far off it is in practice.

**Comment by mikaylagawarecki** [comment]
In terms of performance, a benchmark with the following script where `slow_cross_entropy_loss` is `nn.functional.cross_entropy` and `fast_cross_entropy_loss` is from the repo. The fast kernel seems to perform better on larger shapes. Still have to verify the VRAM usage.


```python
import torch
import torch.utils.benchmark as benchmark
import itertools

def benchmark_torch_function_in_microseconds(f, *args, **kwargs):
    t0 = benchmark.Timer(
        stmt="f(*args, **kwargs)", globals={"args": args, "kwargs": kwargs, "f": f}
    )
    return t0.blocked_autorange().mean * 1e6



def f(ce, inp, targ):
    out = ce(inp, targ)
    res = out.sum().backward()


if __name__ == "__main__":
    torch.set_default_device('cuda')
    Bs = [1, 128, 512]
    Cs = [1, 128, 512]
    Vs = [1, 128, 512]
    for B, C, V in itertools.product(Bs, Cs, Vs):
        inp_fast, targ_fast = torch.randn(B, C, V, requires_grad=True), torch.randint(0, C, (B, V), dtype=torch.long)
        inp_slow, targ_slow = inp_fast.clone().detach().requires_grad_(True), targ_fast.clone().detach()
        nn_ce_fw_bw_time = benchmark_torch_function_in_microseconds(f, slow_cross_entropy_loss, inp_slow, targ_slow)
        fast_ce_fw_bw_time = benchmark_torch_function_in_microseconds(f, fast_cross_entropy_loss, inp_fast.transpose(1, 2).contiguous(), targ_fast)
        print(f"{B}, {C}, {V} | {nn_ce_fw_bw_time} | {fast_ce_fw_bw_time}")
```

The fast kernel seems to perform better on larger shapes


|B, C, V| fast | slow|
|------|-----|-----|
|1, 1, 1 | 235.7435915619135 | 686.2712893635035|
|1, 1, 128 | 268.07614229619503 | 760.502714663744|
|1, 1, 512 | 248.24433587491512 | 669.7270087897778|
|1, 128, 1 | 248.40889126062393 | 773.9283218979836|
|1, 128, 128 | 240.73941446840763 | 634.1504119336605|
|1, 128, 512 | 232.16360807418823 | 638.4236626327038|
|1, 512, 1 | 242.0982364565134 | 664.6045316010714|
|1, 512, 128 | 503.13411466777325 | 675.5183860659599|
|1, 512, 512 | 521.0937410593033 | 645.5060597509146|
|128, 1, 1 | 236.65301129221916 | 675.1588992774487|
|128, 1, 128 | 233.5704043507576 | 685.9271824359894|
|128, 1, 512 | 259.5357205718756 | 753.6098528653383|
|128, 128, 1 | 261.5032531321049 | 724.104467779398|
|128, 128, 128 | 239.3047083169222 | 712.6245573163033|
|128, 128, 512 | 549.420952796936 | 751.8626227974892|
|128, 512, 1 | 247.7816864848137 | 694.0441783517599|
|128, 512, 128 | 1468.988647684455 | 627.6966836303473|
|128, 512, 512 | 2096.496392041445 | 840.0157000869513|
|512, 1, 1 | 238.08901198208332 | 654.1860345751047|
|512, 1, 128 | 235.62197387218475 | 826.1623736470938|
|512, 1, 512 | 269.15206015110016 | 675.908911973238|
|512, 128, 1 | 227.6756875216961 | 673.624349758029|
|512, 128, 128 | 541.7250655591488 | 648.9405147731304|
|512, 128, 512 | 1554.0191531181333 | 1082.6328583061695|
|512, 512, 1 | 265.87958075106144 | 738.2904049009085|
|512, 512, 128 | 2042.4363762140272 | 825.7313948124647|

**Comment by mikaylagawarecki** [comment]
Removing triage review as discussed at nn triage

**Comment by JubilantJerry** [comment]
> The aforementioned kernel seems to be specialized for 3D `input`. In terms of performance, I ran a benchmark with the following script where `slow_cross_entropy_loss` is `nn.functional.cross_entropy` and `fast_cross_entropy_loss` is from the repo. The fast kernel seems to perform better on certain shapes, but not across all shapes. Still have to verify the VRAM usage.
> 
> ```python
> import torch
> import torch.utils.benchmark as benchmark
> import itertools
> 
> def benchmark_torch_function_in_microseconds(f, *args, **kwargs):
>     t0 = benchmark.Timer(
>         stmt="f(*args, **kwargs)", globals={"args": args, "kwargs": kwargs, "f": f}
>     )
>     return t0.blocked_autorange().mean * 1e6
> 
> 
> 
> def f(ce, inp, targ):
>     out = ce(inp, targ)
>     res = out.sum().backward()
> 
> 
> if __name__ == "__main__":
>     torch.set_default_device('cuda')
>     Bs = [1, 128, 512]
>     Cs = [1, 128, 512]
>     Vs = [1, 128, 512]
>     for B, C, V in itertools.product(Bs, Cs, Vs):
>         inp_fast, targ_fast = torch.randn(B, C, V, requires_grad=True), torch.randint(0, C, (B, V), dtype=torch.long)
>         inp_slow, targ_slow = inp_fast.clone().detach().requires_grad_(True), targ_fast.clone().detach()
>         nn_ce_fw_bw_time = benchmark_torch_function_in_microseconds(f, slow_cross_entropy_loss, inp_slow, targ_slow)
>         fast_ce_fw_bw_time = benchmark_torch_function_in_microseconds(f, fast_cross_entropy_loss, inp_fast.transpose(1, 2).contiguous(), targ_fast)
>         print(f"{B}, {C}, {V} | {nn_ce_fw_bw_time} | {fast_ce_fw_bw_time}")
> ```
> 
> Results

Wondering if there any recent results about the VRAM usage of PyTorch vs this Triton kernel? It has been some time since the release of Unsloth, and the VRAM improvements of the free version seem to be very credible by now, being reproduced by many users. According to [here](https://unsloth.ai/blog/mistral-benchmark), a significant portion of the VRAM gains comes from the custom implementation of CrossEntropyLoss.

**Comment by vadimkantorov** [comment]
Curious, does `F.cross_entropy` do the good job when being passed transposed tensor (to ensure that classes are at dim1)? (e.g. would the op do itself the trick of treating all physically leading dimensions (could getting a dimension permutation by stride argsort and then undoing the permutation if needed by `reduction="none"`) as batch dimensions without actually doing contig first?)
